### PR TITLE
/libros/node_handle: keep shared pointer to TopicManager and ServiceM…

### LINKED ISSUE #838: Exception boost::lock_error thrown from shutdown method
We see crashes like this as multi-threaded processes like `rqt_gui` shut down:

```
../../../../../ros-comm-c9e259436a855db5e5f50f3efe2cec60d1f5786e/clients/roscpp/src/libros/init.cpp in ros::shutdown at line 596

../../../../../ros-comm-c9e259436a855db5e5f50f3efe2cec60d1f5786e/clients/roscpp/src/libros/topic_manager.cpp in ros::TopicManager::shutdown at line 93

/usr/include/boost/thread/lock_types.hpp in unique_lock at line 124

/usr/include/boost/thread/lock_types.hpp in boost::unique_lock<boost::mutex>::lock at line 346

/usr/include/boost/thread/pthread/mutex.hpp in lock at line 116

/usr/include/boost/throw_exception.hpp in boost::throw_exception<boost::lock_error> at line 67

/usr/lib/x86_64-linux-gnu/libstdc++.so.6 in __cxa_throw ()

/usr/lib/x86_64-linux-gnu/libstdc++.so.6 in std::terminate() ()
```

This is the code:

```
void TopicManager::shutdown()
{
  boost::mutex::scoped_lock shutdown_lock(shutting_down_mutex_);
  if (shutting_down_)
  {
    return;
  }

  {
    boost::recursive_mutex::scoped_lock lock1(advertised_topics_mutex_);
    boost::mutex::scoped_lock lock2(subs_mutex_);
    shutting_down_ = true;
  }

  ...
```

It looks like the expected behaviour is to block until the mutex can be acquired, so this is maybe an issue with one thread tearing down the TopicManager and then other finding a mutex which is no longer valid? Maybe there's a better solution here than just wrapping a try block around the whole thing?

### PR DESCRIPTION
**Pull request by cwecht**
…anager to avoid boost::lock_error exceptions.

Fixes #838.
This solution does not change the object-layout of NodeHandle.

### CODE REVIEW

**Comment by cwecht** [comment]
Well, this failed because of the flakyness of topic_tools throttle.test ...

File: clients/roscpp/src/libros/node_handle.cpp [hunk 1]
```diff
@@ -55,6 +55,10 @@ boost::mutex g_nh_refcount_mutex;
 int32_t g_nh_refcount = 0;
 bool g_node_started_by_nh = false;
 
+// postpone desctruction of the managers until every NodeHandle has been desctructed.
+TopicManagerPtr g_topic_manager_;
```
Comment (scott-eddy) [inline clients/roscpp/src/libros/node_handle.cpp]: I'm guessing we want these explicitly initialized to NULL

**Comment by scott-eddy** [comment]
This looks good to me

**Comment by cwecht** [comment]
Alternatively, the shared_ptrs could be added NodeHandleBackingCollection. This would also avoid an API break. Since order of destruction of global object can be an issue, this might be a better solution.

**Comment by cwecht** [comment]
The above mentioned alternative is astonishingly simple: https://github.com/cwecht/ros_comm/commit/39b9938dc28a4fddcea9ad3772a6d9d9ff08954f

This might actually be a better solution. The solution of this PR relies (probably?) on specific destruction order of globals/statics, which is unfortunate. The alternative version ensures, that the managers live at least as long as any NodeHandle-instance.

**Comment by scott-eddy** [comment]
@cwecht can you close this in favor of #1656, so it doesn't take up maintainers bandwidth?

**Pull request closed without merging**